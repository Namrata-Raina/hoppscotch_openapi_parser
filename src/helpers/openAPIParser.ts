import { OpenAPIV3 } from "openapi-types"
import { isRef } from "vue"

let components : OpenAPIV3.ComponentsObject = {}

function isOperationType(object: any): object is OpenAPIV3.OperationObject {
  return true
}

function isRequestBodyType(object: any): object is OpenAPIV3.RequestBodyObject {
  return true
}

function isParameterType(object: any): object is OpenAPIV3.ParameterObject {
  return true
}

const isSchemaObjectType = (object: any) : object is OpenAPIV3.SchemaObject => true
const isResponseObjectType = (object: any) : object is OpenAPIV3.ResponseObject => true
const isExampleObjectType = (object: any) : object is OpenAPIV3.ExampleObject => true
const isParameterObjectType = (object: any) : object is OpenAPIV3.ParameterObject => true
const isRequestBodyObjectType = (object: any) : object is OpenAPIV3.RequestBodyObject => true
const isHeaderObjectType = (object: any) : object is OpenAPIV3.HeaderObject => true
const isSecuritySchemeObjectType = (object: any) : object is OpenAPIV3.SecuritySchemeObject => true
const isLinkObjectType = (object: any) : object is OpenAPIV3.LinkObject => true
const isCallbackObjectType = (object: any) : object is OpenAPIV3.CallbackObject => true


const isReferenceObjectType = (object: any) : object is OpenAPIV3.ReferenceObject => true

function fetchComponent(
  child : "schemas" | "responses" | "parameters" | "examples" | "requestBodies" | "headers" | "securitySchemes" | "links" | "callbacks",
  name : string
) {
  const temp: {[key: string] : any} | undefined = components[child];

  if (typeof temp === "undefined") return null;

  let component : any = temp[name];

  if(isReferenceObjectType(component)) {
    component = fetchComponent(child, component["$ref"]?.split('/').pop() ?? "")
  }

  switch(child) {
    case 'schemas':
      component = isSchemaObjectType(component) ? component : null;
      break;

    case 'responses':
      component = isResponseObjectType(component) ? component : null;
      break;

    case 'parameters':
      component = isParameterObjectType(component) ? component : null;
      break;

    case 'examples':
      component = isExampleObjectType(component) ? component : null;
      break;

    case 'requestBodies':
      component = isRequestBodyObjectType(component) ? component : null;
      break;

    case 'headers':
      component = isHeaderObjectType(component) ? component : null;
      break;

    case 'securitySchemes':
      component = isSecuritySchemeObjectType(component) ? component : null;
      break;

    case 'links':
      component = isLinkObjectType(component) ? component : null;
      break;

    case 'callbacks':
      component = isCallbackObjectType(component) ? component : null;
      break;

    default:
  }

  return component;
}

// fetchComponent("securitySchemes", "petstore_auth");

const parsePathItem = function (
  path: String,
  pathItem: OpenAPIV3.PathItemObject
) {
  const parsed: Array<any> = []

  for (const [operationType, operation] of Object.entries(pathItem)) { // for ( const value in Object.values(pathItem) )
    // "/allUsers": { get: {} }
    if (!isOperationType(operation)) continue
    const temp = parseOperation(path, operationType, operation)
    console.log({ temp }) 
    parsed.push(temp)
  }

  return parsed
}

const parseOperation = function (
  path: String,
  operationType: String,
  operation: OpenAPIV3.OperationObject
) : any {
  let requestObject : any = {};
  const values : Array<Object> = [
    parsePath(path),
    parseContentType(operation.requestBody),
    parseMethod(operationType),
    parseSummary(operation.summary),
    parseParams(operation.parameters)
  ];

  let delta : Object = {}
  for( delta in values )
  requestObject = { ...requestObject, ...delta }
  
	return requestObject;
} 

const parseContentType = function (
  requestBody : OpenAPIV3.RequestBodyObject | OpenAPIV3.ReferenceObject | undefined
) {
  if (typeof requestBody === "undefined") return { "contentType" : "" };
  if (isReferenceObjectType(requestBody)) {
    requestBody = fetchComponent("requestBodies" , requestBody["$ref"]?.split('/').pop() ?? "")
  }
  if (isRequestBodyObjectType(requestBody))
    return { "contentType" : Object.keys(requestBody.content)[0] };

  return { "contentType" : "" };
}

const parsePath = function (
  path: String
) {
  return { "path" : path }
}

const parseMethod = function (
  operationType: String
) {
  return { "method" : operationType.toUpperCase() }
}

const parseSummary = function (
  summary: String | undefined
) {
  if(typeof summary !== 'string') return {}
  return { "openapi_summary" : summary }
}

const parseParams = function (
  parameters: (OpenAPIV3.ParameterObject | OpenAPIV3.ReferenceObject)[] | undefined
) {
  if (typeof parameters === 'undefined') return {}
  parameters = parameters.map((p) => {
    if(isReferenceObjectType(p)){
      p = fetchComponent("parameters" , p["$ref"]?.split('/').pop() ?? "")
    }
    return p
  })

  const result = parameters
    .filter((p) : p is OpenAPIV3.ParameterObject => true)
    .filter(p => p.in === 'query')
    .map(p => {
      const temp = {
        key : p.name,
        active : true,
        value: ""
      }
      
      if (p.example) {
        temp.value = p.example
      }
      
      if (p.examples) {
        const f = Object.values(p.examples)
        if (f.length) {
          let temp1 = f[0];
          if(isReferenceObjectType(temp1)){
            temp1 = fetchComponent("examples", temp1["$ref"]?.split('/').pop() ?? "")
          }
          temp.value = isExampleObjectType(temp1) ? temp1.value : ""
        }
      }
      return temp
    });
  return { params : result }
}

const parameters  ={
  examples : {
    "kutteKaNaam" : {/* Reference  Object */},
  }
}

Object.values(parameters.examples)
// [ {/* Reference  Object */}, {/* Example  Object */}]

// parameterObject.examples = {
//  [key: string] : ExampleObject | ReferenceObject
//}

// f(x) =  x > 10    : x - 10
//         otherwise : x

// g(x) = x * 2

// g(f(10))  = 20
// g(f(-10)) = -20



// let obj : Array<any> = [
//   1,
//   2,
//   3,
//   11
// ]

// obj = obj.map(
//   (p) => {

//     if( p > 10 )
//     {
//       return p - 10;
//     }

//     return p;

//   }
// );

// obj = obj.map( 
//   function (val) {
//     if (val <= 10)
//     return "Too small";

//     return "a" + val;
//   }
// )
// => [ "Too small", "Too small", "Too small", "a11" ]



/*
**
    {
      securitySchemes: {
        "blabla" : {
          "type": "http",
          "scheme": "bearer",
          "bearerFormat": "JWT"
        },
        "nonblabla": {
          "type": "http",
          "scheme": "oauth",
          ""
        }
      }
    }
    {
      security: [
        {
          "blabla": [],
        },
        {
          "nonblabla" : ["blablablablabla"]
        }
      ]
    }
**
*/


/*
** const res = { name : "Namrata" }
** res.name == 'Namrata'
*/

export default function (filecontent: OpenAPIV3.Document) {
  const paths = filecontent.paths
  components = filecontent.components ?? {}
  const requests = []
  for (const [path, pathItem] of Object.entries(paths)) {    
    const temp = parsePathItem(path, pathItem ?? {})
    requests.push(...temp)
    // temp = [1, 2, 3]
    // request.push(1, 2, 3) <---> request.push([ 1, 2, 3 ])
  }
  console.log({ requests })
  return requests
}

// let k = {
//     a: {},
//     b: "two"
// }

// Object.keys(k) => ["a", "b"]
// Object.values(k) => [{}, "two"]
// Object.entries(k) => [["a", "one"], ["b", "two"]]

// [
//   {
//     "folders": [
//       {
//         "requests": [],
//         "name": null,
//         "folders": []
//       }
//     ],
//     "name": "My Collection",
//     "requests": [
//       {
//         "url": "https://mail.google.com",
//         "path": "/getAllUsers",
//         "method": "POST",
//         "auth": "None",
//         "httpUser": "",
//         "httpPassword": "",
//         "passwordFieldType": "password",
//         "bearerToken": "required",
//         "bearerRequired": true,
//         "headers": [],
//         "params": [
//           {
//             "key": "user",
//             "value": "",
//             "active": true
//           }
//         ],
//         "bodyParams": [
//           {
//             "active": true,
//             "key": "person",
//             "value": ""
//           },
//           {
//             "key": "",
//             "value": "",
//             "active": true
//           }
//         ],
//         "rawParams": "{\n  \"person\": \"\"\n}",
//         "rawInput": false,
//         "contentType": "application/x-www-form-urlencoded",
//         "requestType": "curl",
//         "preRequestScript": "// pw.env.set('variable', 'value');",
//         "testScript": "// pw.expect('variable').toBe('value');",
//         "name": "blah blah"
//       }
//     ]
//   }
// ]