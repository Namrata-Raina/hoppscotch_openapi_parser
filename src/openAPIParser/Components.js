import { isArray, isObject } from "lodash";
import { isComponentsType, isReferenceObjectType } from "./Validators";
export class Components {
    components;
    constructor(comp) {
        this.components = comp;
    }
    fetchComponent(child, name) {
        const temp = this.components[child];
        if (typeof temp === "undefined")
            return null;
        let component = temp[name];
        while (isReferenceObjectType(component)) {
            component = this.fetchComponent(child, component.$ref?.split("/").pop() ?? "");
        }
        return this.resolveNested(component);
    }
    resolveNested(data) {
        if (isArray(data)) {
            return data.map(data => this.resolveNested(data));
        }
        if (["string", "number", "bigint", "boolean", "undefined", "symbol", "null"].includes(typeof data)) {
            return data;
        }
        if (isReferenceObjectType(data)) {
            const parts = data["$ref"].split("/");
            if (parts[1] == "components" && isComponentsType(parts[2])) {
                const ret = this.resolveNested(this.fetchComponent(parts[2], parts.pop() ?? ""));
                return parts[2] == "examples" ? ret.value : ret;
            }
        }
        if (isObject(data)) {
            return Object.fromEntries(Object.entries(data).map(([key, value]) => [key, this.resolveNested(value)]));
        }
        return data;
    }
}
//# sourceMappingURL=Components.js.map