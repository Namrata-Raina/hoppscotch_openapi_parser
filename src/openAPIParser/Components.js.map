{"version":3,"file":"Components.js","sourceRoot":"","sources":["Components.ts"],"names":[],"mappings":"AACA,OAAO,EAAE,OAAO,EAAE,QAAQ,EAAU,MAAM,QAAQ,CAAA;AAClD,OAAO,EAAiB,gBAAgB,EAAE,qBAAqB,EAAkB,MAAM,cAAc,CAAA;AAErG,MAAM,OAAO,UAAU;IACnB,UAAU,CAA6B;IAEvC,YAAY,IAAgC;QACxC,IAAI,CAAC,UAAU,GAAG,IAAI,CAAA;IAC1B,CAAC;IAED,cAAc,CACV,KAAqB,EACrB,IAAY;QAEZ,MAAM,IAAI,GAAuC,IAAI,CAAC,UAAU,CAAC,KAAK,CAAC,CAAA;QAEvE,IAAI,OAAO,IAAI,KAAK,WAAW;YAAE,OAAO,IAAI,CAAA;QAE5C,IAAI,SAAS,GAAQ,IAAI,CAAC,IAAI,CAAC,CAAA;QAE/B,OAAO,qBAAqB,CAAC,SAAS,CAAC,EAAE;YACvC,SAAS,GAAG,IAAI,CAAC,cAAc,CAAC,KAAK,EAAE,SAAS,CAAC,IAAI,EAAE,KAAK,CAAC,GAAG,CAAC,CAAC,GAAG,EAAE,IAAI,EAAE,CAAC,CAAA;SAC/E;QAED,OAAO,IAAI,CAAC,aAAa,CAAC,SAAS,CAAC,CAAA;IACtC,CAAC;IAED,aAAa,CAAC,IAAqD;QACjE,IAAI,OAAO,CAAC,IAAI,CAAC,EAAE;YACjB,OAAO,IAAI,CAAC,GAAG,CAAC,IAAI,CAAC,EAAE,CAAC,IAAI,CAAC,aAAa,CAAC,IAAI,CAAC,CAAC,CAAA;SAClD;QACD,IAAI,CAAC,QAAQ,EAAE,QAAQ,EAAE,QAAQ,EAAE,SAAS,EAAE,WAAW,EAAE,QAAQ,EAAE,MAAM,CAAC,CAAC,QAAQ,CAAC,OAAO,IAAI,CAAC,EAAE;YAClG,OAAO,IAAI,CAAA;SACZ;QACD,IAAI,qBAAqB,CAAC,IAAI,CAAC,EAAE;YAC/B,MAAM,KAAK,GAAG,IAAI,CAAC,MAAM,CAAC,CAAC,KAAK,CAAC,GAAG,CAAC,CAAA;YACrC,IAAI,KAAK,CAAC,CAAC,CAAC,IAAI,YAAY,IAAI,gBAAgB,CAAC,KAAK,CAAC,CAAC,CAAC,CAAC,EAAE;gBAC1D,MAAM,GAAG,GAAG,IAAI,CAAC,aAAa,CAAC,IAAI,CAAC,cAAc,CAAC,KAAK,CAAC,CAAC,CAAC,EAAE,KAAK,CAAC,GAAG,EAAE,IAAI,EAAE,CAAC,CAAC,CAAA;gBAChF,OAAO,KAAK,CAAC,CAAC,CAAC,IAAI,UAAU,CAAC,CAAC,CAAC,GAAG,CAAC,KAAK,CAAC,CAAC,CAAC,GAAG,CAAA;aAChD;SACF;QACD,IAAI,QAAQ,CAAC,IAAI,CAAC,EAAE;YAClB,OAAO,MAAM,CAAC,WAAW,CAAC,MAAM,CAAC,OAAO,CAAC,IAAI,CAAC,CAAC,GAAG,CAAC,CAAC,CAAC,GAAG,EAAE,KAAK,CAAC,EAAE,EAAE,CAAC,CAAC,GAAG,EAAE,IAAI,CAAC,aAAa,CAAC,KAAK,CAAC,CAAC,CAAC,CAAC,CAAA;SACxG;QACD,OAAO,IAAI,CAAC;IACd,CAAC;CACN"}