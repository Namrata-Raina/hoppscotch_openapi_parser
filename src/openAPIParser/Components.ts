import { OpenAPIV3 } from "openapi-types"
import { isArray, isObject, values } from "lodash"
import {ComponentTypes, isComponentsType, isReferenceObjectType, isOperationType} from "./Validators"

export class Components {
    components : OpenAPIV3.ComponentsObject

    constructor(comp: OpenAPIV3.ComponentsObject) {
        this.components = comp
    }

    fetchComponent(
        child: ComponentTypes,
        name: string
      ) {
        const temp: { [key: string]: any } | undefined = this.components[child]
      
        if (typeof temp === "undefined") return null
      
        let component: any = temp[name]
      
        while (isReferenceObjectType(component)) {
          component = this.fetchComponent(child, component.$ref?.split("/").pop() ?? "")
        }
      
        return this.resolveNested(component)
      }

      resolveNested(data: Object | Array<any> | OpenAPIV3.ReferenceObject) : any {
        if (isArray(data)) {
          return data.map(data => this.resolveNested(data))
        }
        if (["string", "number", "bigint", "boolean", "undefined", "symbol", "null"].includes(typeof data)) {
          return data
        }
        if (isReferenceObjectType(data)) {
          const parts = data["$ref"].split("/")
          if (parts[1] == "components" && isComponentsType(parts[2])) {
            const ret = this.resolveNested(this.fetchComponent(parts[2], parts.pop() ?? ""))
            return parts[2] == "examples" ? ret.value : ret
          }
        }
        if (isObject(data)) {
          return Object.fromEntries(Object.entries(data).map(([key, value]) => [key, this.resolveNested(value)]))
        }
        return data;
      }
}