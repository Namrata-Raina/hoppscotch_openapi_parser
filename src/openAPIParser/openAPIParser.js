export {};
//# sourceMappingURL=openAPIParser.js.map