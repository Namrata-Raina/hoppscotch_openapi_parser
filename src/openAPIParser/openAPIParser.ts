import { JSONSchema4 } from "json-schema"

export type ParamLocations = "query" | "path"
export type ParamTypes = string | number | Object
export type ValidContentTypes = string // TODO: Define Content Types properly

export interface ParsedParameter {
    in: ParamLocations
    key: string
    value: any // Example value, if present. Defaults to "zero" value for type if absent
    type: string
    description?: string
}

export interface ParsedHeader {
    key: string
    value: any // Example value, if present. Defaults to "zero" value for type if absent
    description?: String
}

export interface BodyExample {
    contentType: ValidContentTypes
    summary: string
    value: string // Raw response body
}

export interface RequestBody {
    contentType: ValidContentTypes
    schema?: SchemaObject
    examples?: BodyExample[]
}

export interface ResponseDetails {
    contentType: ValidContentTypes
    schema?: SchemaObject
    examples?: BodyExample[]
}

export interface OpenAPIParsedRequest {
    name: string
    method: string
    path: string // endpoint = hosturl + path
    params: ParsedParameter[]
    headers: ParsedHeader[]
    auth: any // Details are available. Figuring out appropriate structure for representation

    body?: RequestBody
    summary: string
    responses?: BodyExample[] // examples if present
}

export interface OpenAPIParsedSpec {
    name: string
    version: string
    description?: string
    requests: OpenAPIParsedRequest[]
}

/*
    The OpenAPI Schemas follow the JSON Schema Definition [ https://datatracker.ietf.org/doc/html/draft-wright-json-schema-00 ] very closely and extends it
    As such, it is not being parsed into a different format, apart from resolving the references, keeping in mind the standardization. The Schema being used is given below:
*/
export interface SchemaObject extends JSONSchema4 {}
