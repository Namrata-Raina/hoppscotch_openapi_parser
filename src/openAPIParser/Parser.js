import { Components } from "./Components";
import { Schema } from "./Schema";
import { isExampleObjectType, isReferenceObjectType, isOperationType, isRequestBodyObjectType, isParamlocations, isMediaTypeObject } from "./Validators";
export class OpenAPIParser {
    source;
    componentsHolder;
    schemaRef;
    constructor(contents) {
        this.source = contents;
        this.componentsHolder = new Components(this.source.components || {});
        this.schemaRef = new Schema(this.componentsHolder);
    }
    parsePathItem(path, pathItem) {
        const parsed = [];
        for (const [operationType, operation] of Object.entries(pathItem)) {
            if (!isOperationType(operation))
                continue;
            const temp = this.parseOperation(path, operationType, operation);
            parsed.push(temp);
        }
        return parsed;
    }
    parseOperation(path, operationType, operation) {
        const body = this.parseBody(operation.requestBody || undefined);
        const requestObject = {
            name: this.parseName(operation, path, operationType),
            method: this.parseMethod(operationType),
            path: this.parsePath(path),
            params: this.parseParams(operation.parameters),
            headers: this.parseHeaders(operation.parameters),
            auth: undefined,
            body: undefined,
            summary: this.parseSummary(operation.summary)
        };
        if (typeof (body) != "boolean")
            requestObject.body = body;
        return requestObject;
    }
    parseBody(requestBody) {
        if (requestBody === undefined)
            return false;
        if (isReferenceObjectType(requestBody)) {
            requestBody = this.componentsHolder.fetchComponent("requestBodies", requestBody.$ref?.split("/").pop() ?? "");
        }
        if (!isRequestBodyObjectType(requestBody))
            return false;
        let cType = "application/json";
        let def = requestBody.content["application/json"];
        if (!isMediaTypeObject(def)) {
            const meds = Object.entries(requestBody.content);
            if (meds.length === 0)
                return false;
            def = meds[0][1];
            cType = meds[0][0];
        }
        let schema = {};
        const examples = [];
        if (typeof (def.schema) !== 'undefined') {
            let tempSchema = {};
            if (isReferenceObjectType(def.schema)) {
                tempSchema = this.componentsHolder.fetchComponent("schemas", def.schema.$ref?.split("/").pop() ?? "");
            }
            else {
                tempSchema = this.componentsHolder.resolveNested(def.schema);
            }
            schema = this.schemaRef.getJSONSchema(tempSchema);
            const exampleObj = {
                contentType: cType,
                summary: "Generated",
                value: JSON.stringify(this.schemaRef.getJSONExample(schema))
            };
            examples.push(exampleObj);
        }
        if (typeof (def.example) !== "undefined") {
            let tempExample = {};
            if (isReferenceObjectType(def.example)) {
                tempExample = this.componentsHolder.fetchComponent("examples", def.example.$ref?.split("/").pop() ?? "");
            }
            else {
                tempExample = this.componentsHolder.resolveNested(def.example);
            }
            const exampleObj = {
                contentType: cType,
                summary: tempExample.summary || "Imported",
                value: JSON.stringify(tempExample.value || {})
            };
            examples.unshift(exampleObj);
        }
        else if (typeof (def.examples) !== "undefined") {
            Object.values(def.examples).forEach(example => {
                let tempExample = {};
                if (isReferenceObjectType(example)) {
                    tempExample = this.componentsHolder.fetchComponent("examples", example.$ref?.split("/").pop() ?? "");
                }
                else {
                    tempExample = this.componentsHolder.resolveNested(example);
                }
                const exampleObj = {
                    contentType: cType,
                    summary: tempExample.summary || "Imported",
                    value: JSON.stringify(tempExample.value || {})
                };
                examples.unshift(exampleObj);
            });
        }
        return {
            contentType: cType,
            schema,
            examples
        };
    }
    parsePath(path) {
        const pathValue = path.replace(/{([^}]+)}/g, "<<$1>>");
        return pathValue;
    }
    parseMethod(operationType) {
        return operationType.toUpperCase();
    }
    parseSummary(summary) {
        if (typeof summary !== "string")
            return "";
        return summary;
    }
    parseName(operation, path, operationType) {
        let namer = "";
        if (operation.operationId) {
            namer = operation.operationId;
        }
        else {
            namer = `${path} - ${operationType}`;
        }
        return namer;
    }
    parseParams(parameters) {
        if (typeof parameters === "undefined")
            return [];
        parameters = parameters.map((p) => {
            if (isReferenceObjectType(p)) {
                p = this.componentsHolder.fetchComponent("parameters", p.$ref?.split("/").pop() ?? "");
            }
            return p;
        });
        const result = parameters
            .filter((p) => true)
            .filter((p) => isParamlocations(p.in))
            .map((p) => {
            const temp = {
                key: p.name,
                in: isParamlocations(p.in) ? p.in : "query",
                value: "",
                type: "",
                description: p.description
            };
            if (p.in === "path") {
                temp.type = "path";
            }
            if (p.example) {
                temp.value = p.example;
            }
            if (p.examples) {
                const f = Object.values(p.examples);
                if (f.length) {
                    let temp1 = f[0];
                    if (isReferenceObjectType(temp1)) {
                        temp1 = this.componentsHolder.fetchComponent("examples", temp1.$ref?.split("/").pop() ?? "");
                    }
                    temp.value = isExampleObjectType(temp1) ? temp1.value : "";
                }
            }
            return temp;
        });
        return result;
    }
    parseHeaders(parameters) {
        if (typeof parameters === "undefined")
            return [];
        parameters = parameters.map((p) => {
            if (isReferenceObjectType(p)) {
                p = this.componentsHolder.fetchComponent("parameters", p.$ref?.split("/").pop() ?? "");
            }
            return p;
        });
        const result = parameters
            .filter((p) => true)
            .filter((p) => p.in === "header")
            .map((p) => {
            const temp = {
                key: p.name,
                value: "",
                description: p.description // TODO: Assign this properly
            };
            return temp;
        });
        return result;
    }
    parse() {
        const paths = this.source.paths;
        const requests = [];
        for (const [path, pathItem] of Object.entries(paths)) {
            const temp = this.parsePathItem(path, pathItem ?? {});
            requests.push(...temp);
        }
        console.log({ requests });
        return {
            name: this.source.info.title,
            requests,
            version: this.source.info.version,
        };
    }
}
//# sourceMappingURL=Parser.js.map