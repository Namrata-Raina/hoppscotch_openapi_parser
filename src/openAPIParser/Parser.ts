import { OpenAPIV3 } from "openapi-types"
import { ParsedHeader, ParsedParameter, OpenAPIParsedRequest, OpenAPIParsedSpec, BodyExample, RequestBody, SchemaObject } from "./openAPIParser"
import { Components } from "./Components"
import { Schema } from "./Schema";
import {isExampleObjectType, isReferenceObjectType, isOperationType, isRequestBodyObjectType, isParamlocations, isMediaTypeObject} from "./Validators"

export class OpenAPIParser {
  source : OpenAPIV3.Document
  componentsHolder : Components
  schemaRef : Schema

  constructor(contents : OpenAPIV3.Document) {
    this.source = contents
    this.componentsHolder = new Components(this.source.components || {})
    this.schemaRef = new Schema(this.componentsHolder)
  }

  parsePathItem (
    path: string,
    pathItem: OpenAPIV3.PathItemObject
  ) : OpenAPIParsedRequest[] {
    const parsed: OpenAPIParsedRequest[] = []

    for (const [operationType, operation] of Object.entries(pathItem)) {
      if (!isOperationType(operation)) continue
      const temp = this.parseOperation(path, operationType, operation)
      parsed.push(temp)
    }

    return parsed
  }

  parseOperation (
    path: string,
    operationType: string,
    operation: OpenAPIV3.OperationObject
  ): OpenAPIParsedRequest {

    const body = this.parseBody(operation.requestBody || undefined)

    const requestObject: OpenAPIParsedRequest = {
      name: this.parseName(operation, path, operationType),
      method: this.parseMethod(operationType),
      path: this.parsePath(path),
      params: this.parseParams(operation.parameters),
      headers: this.parseHeaders(operation.parameters),
      auth: undefined,
      body: undefined,
      summary: this.parseSummary(operation.summary)
    }

    if (typeof(body) != "boolean")
      requestObject.body = body

    return requestObject
  }

  parseBody (
    requestBody:
      | OpenAPIV3.RequestBodyObject
      | OpenAPIV3.ReferenceObject
      | undefined
  ) : RequestBody | boolean {
    if (requestBody === undefined) return false
    if (isReferenceObjectType(requestBody)) {
      requestBody = this.componentsHolder.fetchComponent(
        "requestBodies",
        requestBody.$ref?.split("/").pop() ?? ""
      )
    }
    if (!isRequestBodyObjectType(requestBody))
      return false

    let cType = "application/json"
    let def = requestBody.content["application/json"]
    if(!isMediaTypeObject(def)) {
      const meds = Object.entries(requestBody.content)
      if (meds.length === 0)
        return false
      def = meds[0][1]
      cType = meds[0][0]
    }

    let schema : SchemaObject = {}
    const examples : BodyExample[] = []

    if (typeof(def.schema) !== 'undefined') {
      let tempSchema : OpenAPIV3.SchemaObject = {}
      if (isReferenceObjectType(def.schema)) {
        tempSchema = this.componentsHolder.fetchComponent("schemas", def.schema.$ref?.split("/").pop() ?? "")
      } else {
        tempSchema = this.componentsHolder.resolveNested(def.schema)
      }
      schema = this.schemaRef.getJSONSchema(tempSchema)
      const exampleObj : BodyExample = {
        contentType: cType,
        summary: "Generated",
        value: JSON.stringify(this.schemaRef.getJSONExample(schema))
      }
      examples.push(exampleObj)
    }

    if (typeof(def.example) !== "undefined") {
      let tempExample : OpenAPIV3.ExampleObject = {}
      if (isReferenceObjectType(def.example)) {
        tempExample = this.componentsHolder.fetchComponent("examples", def.example.$ref?.split("/").pop() ?? "")
      } else {
        tempExample = this.componentsHolder.resolveNested(def.example)
      }
      const exampleObj : BodyExample = {
        contentType: cType,
        summary: tempExample.summary || "Imported",
        value: JSON.stringify(tempExample.value || {})
      }
      examples.unshift(exampleObj)
    } else if (typeof(def.examples) !== "undefined") {
      Object.values(def.examples).forEach( example => {
        let tempExample : OpenAPIV3.ExampleObject = {}
        if (isReferenceObjectType(example)) {
          tempExample = this.componentsHolder.fetchComponent("examples", example.$ref?.split("/").pop() ?? "")
        } else {
          tempExample = this.componentsHolder.resolveNested(example)
        }
        const exampleObj : BodyExample = {
          contentType: cType,
          summary: tempExample.summary || "Imported",
          value: JSON.stringify(tempExample.value || {})
        }
        examples.unshift(exampleObj)
      })
    }
    
    return {
      contentType: cType,
      schema,
      examples
    }
  }

  parsePath (path: string) : string {
    const pathValue = path.replace(/{([^}]+)}/g, "<<$1>>")
    return pathValue
  }

  parseMethod (operationType: string) : string {
    return operationType.toUpperCase()
  }

  parseSummary (summary: String | undefined) : string {
    if (typeof summary !== "string") return ""
    return summary
  }

  parseName (
    operation: OpenAPIV3.OperationObject,
    path: String,
    operationType: String
  ) : string {
    let namer = ""
    if (operation.operationId) {
      namer = operation.operationId
    } else {
      namer = `${path} - ${operationType}`
    }
    return namer
  }

  parseParams (
    parameters:
      | (OpenAPIV3.ParameterObject | OpenAPIV3.ReferenceObject)[]
      | undefined
  ) : ParsedParameter[] {
    if (typeof parameters === "undefined") return []
    parameters = parameters.map((p) => {
      if (isReferenceObjectType(p)) {
        p = this.componentsHolder.fetchComponent("parameters", p.$ref?.split("/").pop() ?? "")
      }
      return p
    })

    const result = parameters
      .filter((p): p is OpenAPIV3.ParameterObject => true)
      .filter((p) => isParamlocations(p.in))
      .map((p) => {
        const temp: ParsedParameter = {
          key: p.name,
          in: isParamlocations(p.in) ? p.in : "query",
          value: "",
          type: "", // TODO: Assign this properly
          description: p.description
        }

        if (p.in === "path") {
          temp.type = "path"
        }

        if (p.example) {
          temp.value = p.example
        }

        if (p.examples) {
          const f = Object.values(p.examples)
          if (f.length) {
            let temp1 = f[0]
            if (isReferenceObjectType(temp1)) {
              temp1 = this.componentsHolder.fetchComponent(
                "examples",
                temp1.$ref?.split("/").pop() ?? ""
              )
            }
            temp.value = isExampleObjectType(temp1) ? temp1.value : ""
          }
        }
        return temp
      })
    return result
  }

  parseHeaders (
    parameters:
      | (OpenAPIV3.ParameterObject | OpenAPIV3.ReferenceObject)[]
      | undefined
  ) : ParsedHeader[] {
    if (typeof parameters === "undefined") return []
    parameters = parameters.map((p) => {
      if (isReferenceObjectType(p)) {
        p = this.componentsHolder.fetchComponent("parameters", p.$ref?.split("/").pop() ?? "")
      }
      return p
    })

    const result = parameters
      .filter((p): p is OpenAPIV3.ParameterObject => true)
      .filter((p) => p.in === "header")
      .map((p) => {
        const temp: ParsedHeader = {
          key: p.name,
          value: "",
          description: p.description // TODO: Assign this properly
        }
        return temp
      })
    return result
  }

  parse () : OpenAPIParsedSpec {
    const paths = this.source.paths
    const requests = []
    for (const [path, pathItem] of Object.entries(paths)) {
      const temp = this.parsePathItem(path, pathItem ?? {})
      requests.push(...temp)
    }
    console.log({ requests })
    return {
      name: this.source.info.title,
      requests,
      version: this.source.info.version,
    }
  }
}