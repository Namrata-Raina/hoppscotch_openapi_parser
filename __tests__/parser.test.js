import { OpenAPIParser }  from "../src/openAPIParser/Parser"

import * as apiWithExamples from "./data/api-with-examples.json"
import * as callbackExample from "./data/callback-example.json"
import * as linkExample from "./data/link-example.json"
import * as petstore from "./data/petstore.json"
import * as petstoreExpanded from "./data/petstore-expanded.json"
import * as uspto from "./data/uspto.json"

function determiner(pp) {
    var parsedOb = new OpenAPIParser(pp).parse()
    parsedOb.requests = parsedOb.requests.map(r => ({
        ...r,
        body: r.body && {
            ...r.body,
            examples: undefined
        },
        responses: r.responses && r.responses.map(resp => undefined)
    }))
}

it("api-with-examples", () => {
    expect(determiner(apiWithExamples)).toMatchSnapshot()
});

it("callback-example", () => {
    expect(determiner(callbackExample)).toMatchSnapshot()
});

it("link-example", () => {
    expect(determiner(linkExample)).toMatchSnapshot()
});

it("petstore", () => {
    expect(determiner(petstore)).toMatchSnapshot()
});

it("petstore expanded", () => {
    const parsedPE = determiner(petstoreExpanded)
    expect(parsedPE).toMatchSnapshot()
});

it("upsto", () => {
    var usptoParsed = determiner(uspto)
    expect(usptoParsed).toMatchSnapshot()
});