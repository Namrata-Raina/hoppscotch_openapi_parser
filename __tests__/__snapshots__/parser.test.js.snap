// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`api-with-examples 1`] = `undefined`;

exports[`callback-example 1`] = `undefined`;

exports[`link-example 1`] = `undefined`;

exports[`petstore 1`] = `undefined`;

exports[`petstore expanded 1`] = `undefined`;

exports[`upsto 1`] = `undefined`;
